mathics/builtin/matrices/partmatrix.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+
+"""
+Parts of Matrices
+
+Set of methods for manipulating Matrices.
+"""
+
+
+from mathics.builtin.base import Builtin
+from mathics.builtin.tensors import get_dimensions
+from mathics.core.expression import Expression
+from mathics.core.atoms import Integer
+
+
+class Diagonal(Builtin):
+    """
+    <dl>
+        <dt>'Diagonal[$m$]'
+        <dd>gives a list with the values in the diagonal of the matrix $m$.
+        <dt>'Diagonal[$m$, $k$]'
+        <dd>gives a list with the values in the $k$ diagonal of the matrix $m$.
+    </dl>
+
+    >> Diagonal[{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}]
+     = {1, 5, 9}
+
+
+
+    >> Diagonal[{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}, 1]
+     = {2, 6}
+    """
+
+    rules = {
+        "Diagonal[expr_]": "Diagonal[expr, 0]",
+    }
+
+    summary_text = "Return a list with the diagonal elements of a given matrix"
+
+    def apply(self, expr, diag, evaluation):
+        "Diagonal[expr_List, diag_Integer]"
+
+        result = []
+
+        for count, value in enumerate(expr.elements, start=diag.value):
+            if not hasattr(value, "elements") or len(value.elements) <= count:
+                break
+            if count < 0:
```

Comment:
Why would the index be negative?

Reply:
If you want to get a diagonal lower than the leading one you need to pass a negative `diag`.
It's the same behaviour of Mathematica; the if statement is needed because e.g. `Diagonal[matrix, -1] ` would return a list starting with the last element of the first row